rllib/agents/ppo/ppo_tf_policy.py

```diff
@@ -385,61 +53,204 @@ def setup_config(
         )
 
 
-def setup_mixins(
-    policy: Policy,
-    obs_space: gym.spaces.Space,
-    action_space: gym.spaces.Space,
-    config: TrainerConfigDict,
-) -> None:
-    """Call mixin classes' constructors before Policy's loss initialization.
+# We need this builder function because we want to share the same
+# custom logics between TF1 static and TF2 eager policies.
+def get_ppo_tf_policy(base: type) -> type:
+    """Construct a PPOTFPolicy inheriting either static or eager base policies.
 
     Args:
-        policy (Policy): The Policy object.
-        obs_space (gym.spaces.Space): The Policy's observation space.
-        action_space (gym.spaces.Space): The Policy's action space.
-        config (TrainerConfigDict): The Policy's config.
+        base: Base class for this policy. DynamicTFPolicyV2 or EagerTFPolicyV2.
+
+    Returns:
+        A TF Policy to be used with PPOTrainer.
     """
-    ValueNetworkMixin.__init__(policy, obs_space, action_space, config)
-    KLCoeffMixin.__init__(policy, config)
-    EntropyCoeffSchedule.__init__(
-        policy, config["entropy_coeff"], config["entropy_coeff_schedule"]
-    )
-    LearningRateSchedule.__init__(policy, config["lr"], config["lr_schedule"])
-
-
-@Deprecated(
-    old="rllib.agents.ppo.ppo_tf_policy.postprocess_ppo_gae",
-    new="rllib.evaluation.postprocessing.compute_gae_for_sample_batch",
-    error=False,
-)
-def postprocess_ppo_gae(
-    policy: Policy,
-    sample_batch: SampleBatch,
-    other_agent_batches: Optional[Dict[AgentID, SampleBatch]] = None,
-    episode: Optional[Episode] = None,
-) -> SampleBatch:
-
-    return compute_gae_for_sample_batch(
-        policy, sample_batch, other_agent_batches, episode
-    )
-
-
-# Build a child class of `DynamicTFPolicy`, given the custom functions defined
-# above.
-PPOTFPolicy = build_tf_policy(
-    name="PPOTFPolicy",
-    loss_fn=ppo_surrogate_loss,
-    get_default_config=lambda: ray.rllib.agents.ppo.ppo.DEFAULT_CONFIG,
-    postprocess_fn=compute_gae_for_sample_batch,
-    stats_fn=kl_and_loss_stats,
-    compute_gradients_fn=compute_and_clip_gradients,
-    extra_action_out_fn=vf_preds_fetches,
-    before_init=setup_config,
-    before_loss_init=setup_mixins,
-    mixins=[
-        LearningRateSchedule,
+    class PPOTFPolicy(
         EntropyCoeffSchedule,
+        LearningRateSchedule,
         KLCoeffMixin,
         ValueNetworkMixin,
-    ],
-)
+        base
+    ):
+        def __init__(
+            self,
+            obs_space,
+            action_space,
+            config,
+            existing_model=None,
+            existing_inputs=None,
+        ):
+            validate_config(config)
```

Comment:
should we make this a Policy method as well? Analogous to `Trainer.validate_config()`?

Reply:
It's actually quite different for different algorithms. and some agents don't do this.
we can make it a no-op API call though, that may be useful.
but then the thing is you want to validate this first thing before you call into Policy's __init__().

Reply:
Actually, it's fine. I feel like any config validation should be removed from both Trainers AND policies and completely moved into the new TrainerConfig classes.